File: 0124. Binary Tree Maximum Path Sum/Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.res = float('-inf')
        
    def helper(self, node: TreeNode) -> int:
        if not node:
            return 0;
        
        l = self.helper(node.left)
        r = self.helper(node.right)
        
        self.res = max(self.res, l + r + node.val)
        return max(0, max(l, r) + node.val)
    
    def maxPathSum(self, root: TreeNode) -> int:
        self.helper(root)
        return self.res

File: 0124. Binary Tree Maximum Path Sum/Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int res = Integer.MIN_VALUE;
    
    private int helper(TreeNode node) {
        if (node == null) return 0;
        int l = helper(node.left);
        int r = helper(node.right);
        
        res = Math.max(res, l + r + node.val);
        return Math.max(0, Math.max(l, r) + node.val);
    }
    
    public int maxPathSum(TreeNode root) {
        helper(root);
        return res;
    }
}